wezterm-ssh/src/session/sftp.rs

```diff
@@ -0,0 +1,419 @@
+                path: path.into(),
+                target: target.into(),
+                reply,
+            })))
+            .await?;
+        let result = rx.recv().await?;
+        Ok(result)
+    }
+
+    /// Read a symlink at `path`.
+    ///
+    /// See [`Sftp::readlink`] for more information.
+    pub async fn readlink(&self, path: impl Into<PathBuf>) -> anyhow::Result<PathBuf> {
+        let (reply, rx) = bounded(1);
+        self.tx
+            .send(SessionRequest::Sftp(SftpRequest::Readlink(Readlink {
+                path: path.into(),
+                reply,
+            })))
+            .await?;
+        let result = rx.recv().await?;
+        Ok(result)
+    }
+
+    /// Resolve the real path for `path`.
+    ///
+    /// See [`Sftp::realpath`] for more information.
+    pub async fn realpath(&self, path: impl Into<PathBuf>) -> anyhow::Result<PathBuf> {
+        let (reply, rx) = bounded(1);
+        self.tx
+            .send(SessionRequest::Sftp(SftpRequest::Realpath(Realpath {
+                path: path.into(),
+                reply,
+            })))
+            .await?;
+        let result = rx.recv().await?;
+        Ok(result)
+    }
+
+    /// Rename the filesystem object on the remote filesystem.
+    ///
+    /// See [`Sftp::rename`] for more information.
+    pub async fn rename(
+        &self,
+        src: impl Into<PathBuf>,
+        dst: impl Into<PathBuf>,
+        flags: Option<RenameFlags>,
+    ) -> anyhow::Result<()> {
+        let (reply, rx) = bounded(1);
+        self.tx
+            .send(SessionRequest::Sftp(SftpRequest::Rename(Rename {
+                src: src.into(),
+                dst: dst.into(),
+                flags,
+                reply,
+            })))
+            .await?;
+        let result = rx.recv().await?;
+        Ok(result)
+    }
+
+    /// Remove a file on the remote filesystem.
+    ///
+    /// See [`Sftp::unlink`] for more information.
+    pub async fn unlink(&self, file: impl Into<PathBuf>) -> anyhow::Result<()> {
+        let (reply, rx) = bounded(1);
+        self.tx
+            .send(SessionRequest::Sftp(SftpRequest::Unlink(Unlink {
+                file: file.into(),
+                reply,
+            })))
+            .await?;
+        let result = rx.recv().await?;
+        Ok(result)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum SftpRequest {
+    OpenMode(OpenMode),
+    Open(Open),
+    Create(Create),
+    Opendir(Opendir),
+    Readdir(Readdir),
+    Mkdir(Mkdir),
+    Rmdir(Rmdir),
+    Stat(Stat),
+    Lstat(Lstat),
+    Setstat(Setstat),
+    Symlink(Symlink),
+    Readlink(Readlink),
+    Realpath(Realpath),
+    Rename(Rename),
+    Unlink(Unlink),
+
+    /// Specialized type for file-based operations
+    File(FileRequest),
+}
+
+pub(crate) struct OpenMode {
+    pub filename: PathBuf,
+    pub flags: OpenFlags,
+    pub mode: i32,
+    pub open_type: OpenType,
+    pub reply: Sender<File>,
```

Comment:
re: your question about propagating errors, you could use `Sender<Result<File>>` if the `Result`'s `Err` type is `Send + Sync`.

I would suggest using the `thiserror` crate to define an exportable error variant of our own to represent the significant types of errors that folks would want to handle in code, rather than using the ssh2 crate error types (which are honestly a bit awkward: lots of possible errors in the API, but relatively few of them are used in practice).

Again: a bit of a boilerplatey PITA.

Reply:
Yeah, I think I'll revise to do that. It's really annoying right now on distant as the only error you get is a message about the receiver being closed without clarity about the real problem.

Reply:
Tackled this for SFTP specifically. Out of scope for broader ssh.